agent/plugin/definition.go

```diff
@@ -0,0 +1,154 @@
+package plugin
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/buildkite/agent/yamltojson"
+	"github.com/qri-io/jsonschema"
+)
+
+var ErrDefinitionNotFound = errors.New("Definition file not found")
+
+// Definition defines the plugin.yml file that each plugin has
+type Definition struct {
+	Name          string                 `json:"name"`
+	Requirements  []string               `json:"requirements"`
+	Configuration *jsonschema.RootSchema `json:"configuration"`
+}
+
+// ParseDefinition parses either yaml or json bytes into a Definition
+func ParseDefinition(b []byte) (*Definition, error) {
+	var parsed interface{}
+
+	// Parse the definition as a json compatible string map, the plain yaml
+	// Unmarshal returns a structure that has map[interface{}]interface{}
+	// which causes anything that expects map[string]interface{} to break
+	if err := yamltojson.UnmarshalAsStringMap(b, &parsed); err != nil {
+		return nil, err
+	}
+
+	// Check we've got a map, vs a list or something else, this lets us
+	// give a useful error message
+	parsedMap, ok := parsed.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Failed to convert %T to map[string]interface{}", parsed)
+	}
+
+	// Marshal the whole lot back into json which will let the jsonschema library
+	// parse the schema into and object tree 💃🏼
+	jsonBytes, err := json.Marshal(parsedMap)
+	if err != nil {
+		return nil, err
+	}
+
+	var def Definition
+	if err := json.Unmarshal(jsonBytes, &def); err != nil {
+		return nil, err
+	}
+
+	return &def, nil
+}
+
+// LoadDefinitionFromDir looks in a directory for either a plugin.json or a plugin.yml
+func LoadDefinitionFromDir(dir string) (*Definition, error) {
+	f, err := findDefinitionFile(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := ioutil.ReadFile(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseDefinition(b)
+}
+
+// findDefinitionFile searches for known plugin definition files
+func findDefinitionFile(dir string) (string, error) {
+	var possibleFilenames = []string{
+		filepath.Join(dir, `plugin.json`),
+		filepath.Join(dir, `plugin.yaml`),
+		filepath.Join(dir, `plugin.yml`),
+	}
+	for _, filename := range possibleFilenames {
+		if _, err := os.Stat(filename); err == nil {
+			return filename, nil
+		}
+	}
+	return "", ErrDefinitionNotFound
+}
+
+type Validator struct {
+	CommandExists func(string) bool
```

Comment:
Is this configurable to make testing easier?

Reply:
Yup, for testing purposes.